include/simdjson/inline/document_stream.h

```diff
@@ -136,148 +48,112 @@ really_inline document_stream::iterator::iterator(document_stream& _stream, bool
-      load_next_batch = true;
-      res = simdjson::SUCCESS_AND_HAS_MORE;
-    }
+inline size_t document_stream::next_batch_start() const noexcept {
+  return batch_start + parser.implementation->structural_indexes[parser.implementation->n_structural_indexes];
+}
+
+inline error_code document_stream::run_stage1(dom::parser &p, size_t _batch_start) noexcept {
+  // If this is the final batch, pass partial = false
+  size_t remaining = len - _batch_start;
+  if (remaining <= batch_size) {
+    return p.implementation->stage1(&buf[_batch_start], remaining, false);
+  } else {
+    return p.implementation->stage1(&buf[_batch_start], batch_size, true);
   }
-  return res;
 }
 
-#else  // SIMDJSON_THREADS_ENABLED
+#ifdef SIMDJSON_THREADS_ENABLED
+
+inline void document_stream::load_from_stage1_thread() noexcept {
+  stage1_thread.join();
 
-// single-threaded version of json_parse
-inline error_code document_stream::json_parse() noexcept {
-  error = parser.ensure_capacity(_batch_size);
-  if (error) { return error; }
+  // Swap to the parser that was loaded up in the thread. Make sure the parser has
+  // enough memory to swap to, as well.
+  std::swap(parser, stage1_thread_parser);
+  error = stage1_thread_error;
+  if (error) { return; }
 
-  if (unlikely(load_next_batch)) {
-    advance(current_buffer_loc);
-    n_bytes_parsed += current_buffer_loc;
-    _batch_size = (std::min)(_batch_size, remaining());
-    _batch_size = internal::trimmed_length_safe_utf8((const char *)buf(), _batch_size);
-    auto stage1_is_ok = (error_code)parser.implementation->stage1(buf(), _batch_size, true);
-    if (stage1_is_ok != simdjson::SUCCESS) {
-      return stage1_is_ok;
-    }
-    uint32_t last_index = internal::find_last_json_buf_idx(buf(), _batch_size, parser);
-    if (last_index == 0) {
-      if (parser.implementation->n_structural_indexes == 0) {
-        return EMPTY;
-      }
-    } else {
-      parser.implementation->n_structural_indexes = last_index + 1;
-    }
-    load_next_batch = false;
-  } // load_next_batch
-  error_code res = parser.implementation->stage2(buf(), remaining(), parser.doc, next_json);
-  if (likely(res == simdjson::SUCCESS_AND_HAS_MORE)) {
-    n_parsed_docs++;
-    current_buffer_loc = parser.implementation->structural_indexes[next_json];
-  } else if (res == simdjson::SUCCESS) {
-    n_parsed_docs++;
-    if (remaining() > _batch_size) {
-      current_buffer_loc = parser.implementation->structural_indexes[next_json - 1];
-      next_json = 1;
-      load_next_batch = true;
-      res = simdjson::SUCCESS_AND_HAS_MORE;
-    }
+  // If there's anything left, start the stage 1 thread!
+  if (next_batch_start() < len) {
+    start_stage1_thread();
   }
-  return res;
 }
+
+inline void document_stream::start_stage1_thread() noexcept {
+  // we call the thread on a lambda that will update
+  // this->stage1_thread_error
+  // there is only one thread that may write to this value
+  // TODO this is NOT exception-safe.
```

Comment:
Why don't we make run_stage1 a noexcept? I think that all it does is that it calls `stage1` which is itself noexcept (and thread-safe). This makes it impossible, theoretically, that the thread would have an exception... (that would be bad for obvious reasons).

Reply:
Right, that method *is* exception safe, but that's the concern (I'll clarify the comment). The issue is that, if the std::thread constructor cannot acquire enough system resources for a thread, it will throw an exception. It's not that this thing isn't thread safe, it's just that it's possible it could raise an exception (or more likely abort) when we promised the user that wouldn't happen.

I couldn't find a way to get an error code back from `std::thread()` besides catching the exception.

Reply:
Good point. So there is no way to call std::thread on a function without possibly throwing  `std::system_error`. So why don't we try to catch the exceptions?

Reply:
It might be fine? I don't really know what C++ will do if you put a try catch in a `noexcept` method. But the docs indicate that it may simply abort when it can't get resources, instead.